### PR TITLE
Adding comments to the write path

### PR DESCRIPTION
**Pull request by maysamyabandeh**
also did minor refactoring

### CODE REVIEW

File: db/db_impl_write.cc [hunk 1]
```diff
@@ -341,11 +344,22 @@ Status DBImpl::PreprocessWrite(const WriteOptions& write_options,
   }
 
   if (status.ok() && need_log_sync) {
+    // Wait until the parallel syncs are finished. Any sync process has to sync
+    // the front log too so it is enough to check the status of front()
+    // We do a while loop since log_sync_cv_ is signalled when any sync is finished
+    // Note: there does not seem to be a reason to wait for parallel sync at
```
Comment (maysamyabandeh) [inline db/db_impl_write.cc]: @al13n321 Can you chime in regarding this comment? Thanks

File: db/db_impl_write.cc [hunk 2]
```diff
@@ -341,11 +344,22 @@ Status DBImpl::PreprocessWrite(const WriteOptions& write_options,
...
+    // need_log_sync are usually not used together.
     while (logs_.front().getting_synced) {
       log_sync_cv_.Wait();
     }
     for (auto& log : logs_) {
       assert(!log.getting_synced);
+      // This is just to prevent the logs to be synced by a parallel SyncWAL
+      // call. We will do the actual syncing later after we will write to the
+      // WAL.
+      // Note: there does not seem to be a reason to set this early before we
```
Comment (maysamyabandeh) [inline db/db_impl_write.cc]: @al13n321 also here.

Comment (al13n321) [inline db/db_impl_write.cc]: Iirc, getting_synced is assigned early because it's protected by mutex_, and mutex_ will be unlocked for WAL write and memtable insert. But maybe things changed since then: this code used to be in WriteImpl(), and PreprocessWrite() didn't exist.

File: db/db_impl_write.cc (hunk 1 above)
Comment (al13n321) [inline db/db_impl_write.cc]: I don't remember much, but this seems accurate. Maybe the waiting can be solved by turning getting_synced into a counter instead of bool, but as you said SyncWAL and need_log_sync are usually not used together.

File: db/db_impl_write.cc [hunk 3]
```diff
@@ -145,15 +144,13 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
...
+    // 3. Merges are not okay
     //
-    // Rules 1..3 are enforced by checking the options
+    // Rules 1..2 are enforced by checking the options
     // during startup (CheckConcurrentWritesSupported), so if
     // options.allow_concurrent_memtable_write is true then they can be
-    // assumed to be true.  Rule 4 is checked for each batch.  We could
-    // relax rules 2 and 3 if we could prevent write batches from referring
+    // assumed to be true.  Rule 3 is checked for each batch.  We could
+    // relax rules 2 if we could prevent write batches from referring
```
Comment (yiwu-arbug) [inline db/db_impl_write.cc]: On a side note, I think we can relax rule 3 if max_successive_merge is not enabled. Would be great if some one can confirm.

File: db/db_impl_write.cc [hunk 4]
```diff
@@ -173,6 +170,7 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
       }
     }
 
+    uint64_t last_sequence = versions_->LastSequence();
```
Comment (yiwu-arbug) [inline db/db_impl_write.cc]: Is all the tests pass after this change? (Don't know how to read the phabricator result after the fbcode migjration..)

File: include/rocksdb/options.h [hunk 5]
```diff
@@ -412,6 +412,8 @@ struct DBOptions {
   // This parameter should be set to true while storing data to
   // filesystem like ext3 that can lose files after a reboot.
   // Default: false
+  // Note: on many platforms fdatasync is defined as fsync, so this parameter
```
Comment (yiwu-arbug) [inline include/rocksdb/options.h]: I believe use_data_sync option have been removed?

Comment (maysamyabandeh) [inline include/rocksdb/options.h]: There is currently no use_data_sync in the code base. Can you tell more how would it affect the added note?

Comment (yiwu-arbug) [inline include/rocksdb/options.h]: Nothing. I misread the code and the comment.

File: db/db_impl_write.cc (hunk 4 above)
Comment (maysamyabandeh) [inline db/db_impl_write.cc]: only tsan and asan are timing out which does not seem related.

Comment (yiwu-arbug) [inline db/db_impl_write.cc]: cool.
